### ISSUE TITLE
[Phase 3] Spatial Feature Engineering

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/real-estate-valuation-engine#5]
## Summary

Implement spatial feature engineering using k-nearest neighbors, inverse distance weighting, and geospatial clustering. These features capture location-based value drivers and neighborhood effects.

**Part of:** #1
**Depends on:** #4 (Phase 2)

## Objectives

- [ ] Implement `SpatialFeatureEngine` class
- [ ] Build KD-tree for fast neighbor queries
- [ ] Generate k-NN distance features
- [ ] Generate k-NN price features (mean, std, IDW)
- [ ] Implement HDBSCAN clustering for neighborhoods
- [ ] Add cluster-based features
- [ ] Handle coordinate projection (lat/long to meters)
- [ ] Prevent price leakage from test set
- [ ] Write unit tests

## Files to Modify

| File | Lines | Action | Description |
|------|-------|--------|-------------|
| `src/features/spatial.py` | 1-150 | Implement | SpatialFeatureEngine class |
| `tests/features/test_spatial.py` | NEW | Create | Unit tests |

### SpatialFeatureEngine Class

```python
# src/features/spatial.py
import numpy as np
import pandas as pd
import geopandas as gpd
from scipy.spatial import cKDTree
from sklearn.cluster import HDBSCAN
from typing import List
from dataclasses import dataclass

@dataclass
class SpatialConfig:
    """Configuration for spatial feature engineering."""
    k_neighbors: List[int] = (3, 5, 10, 20)
    idw_power: float = 2.0
    min_cluster_size: int = 10
    use_price_features: bool = True

class SpatialFeatureEngine:
    """Generate spatial features from property locations."""

    def __init__(self, config: SpatialConfig = None):
        self.config = config or SpatialConfig()
        self._tree: cKDTree = None
        self._train_coords: np.ndarray = None
        self._train_prices: np.ndarray = None
        self._cluster_model: HDBSCAN = None

    def fit(self, gdf: gpd.GeoDataFrame, price_col: str = 'sale_price'):
        """Build spatial index from training properties."""
        # Project to meters for distance calculations
        gdf_projected = gdf.to_crs("EPSG:3857")
        self._train_coords = np.column_stack([
            gdf_projected.geometry.x,
            gdf_projected.geometry.y
        ])
        self._train_prices = gdf[price_col].values

        # Build KD-tree
        self._tree = cKDTree(self._train_coords)

        # Fit clustering
        self._cluster_model = HDBSCAN(
            min_cluster_size=self.config.min_cluster_size
        )
        self._cluster_model.fit(self._train_coords)

        return self

    def transform(self, gdf: gpd.GeoDataFrame) -> pd.DataFrame:
        """Generate spatial features for properties."""
        gdf_projected = gdf.to_crs("EPSG:3857")
        coords = np.column_stack([
            gdf_projected.geometry.x,
            gdf_projected.geometry.y
        ])

        features = {}

        for k in self.config.k_neighbors:
            distances, indices = self._tree.query(coords, k=k+1)
            distances = distances[:, 1:]  # Skip self
            indices = indices[:, 1:]

            # Distance features
            features[f'dist_to_{k}nn_mean'] = distances.mean(axis=1)
            features[f'dist_to_{k}nn_min'] = distances.min(axis=1)

            if self.config.use_price_features:
                # Price features from TRAINING data only
                neighbor_prices = self._train_prices[indices]
                features[f'price_{k}nn_mean'] = neighbor_prices.mean(axis=1)
                features[f'price_{k}nn_std'] = neighbor_prices.std(axis=1)

                # IDW-weighted price
                weights = 1 / (distances ** self.config.idw_power + 1e-10)
                weights = weights / weights.sum(axis=1, keepdims=True)
                features[f'price_{k}nn_idw'] = (neighbor_prices * weights).sum(axis=1)

        # Cluster features
        cluster_labels = self._cluster_model.fit_predict(coords)
        features['cluster_id'] = cluster_labels
        features['is_noise'] = (cluster_labels == -1).astype(int)

        return pd.DataFrame(features, index=gdf.index)
```

### Test: No Price Leakage

```python
# tests/features/test_spatial.py
import pytest
import numpy as np
import geopandas as gpd
from shapely.geometry import Point
from src.features.spatial import SpatialFeatureEngine

def test_spatial_features_no_price_leakage():
    """Spatial features should not use test set prices."""
    # Train: low prices
    train_gdf = gpd.GeoDataFrame({
        'sale_price': [100000, 100000, 100000],
        'geometry': [Point(-122.4, 37.7), Point(-122.41, 37.71), Point(-122.42, 37.72)]
    }, crs="EPSG:4326")

    # Test: high prices (should NOT be used)
    test_gdf = gpd.GeoDataFrame({
        'sale_price': [1000000, 1000000],  # 10x higher
        'geometry': [Point(-122.405, 37.705), Point(-122.415, 37.715)]
    }, crs="EPSG:4326")

    engine = SpatialFeatureEngine()
    engine.fit(train_gdf, price_col='sale_price')
    test_features = engine.transform(test_gdf)

    # Neighbor prices should reflect TRAIN (~100k), not test (~1M)
    assert test_features['price_3nn_mean'].max() < 200000
```

## Verification

```bash
pytest tests/features/test_spatial.py -v
```

## Technical Challenges

1. **Coordinate Reference Systems** - Lat/long distances are not linear
   - Solution: Project to EPSG:3857 (meters) for distance calculations
2. **Self-neighbors** - Training points find themselves as nearest
   - Solution: Query k+1 neighbors, skip first
3. **Price leakage** - Test prices must not influence test features
   - Solution: Only store training prices in tree

## Learning Context

> **CRS (Coordinate Reference System):** Geographic coordinates use degrees, but distance calculations need meters. EPSG:4326 is lat/long, EPSG:3857 uses meters. Always convert before distance calculations.

## Definition of Done

- [ ] Spatial features generated without errors
- [ ] Distance units are in meters (not degrees)
- [ ] No price leakage from test set
- [ ] Clustering produces meaningful groups
- [ ] All unit tests pass

## Labels

`phase-3`, `features`, `spatial`, `enhancement`
